src/cmd/cloud.js

```diff
@@ -41,311 +43,381 @@ const PLATFORMS = extend(utilities.knownPlatforms(), {
-			return prompt([{
-				name: 'type',
-				type: 'list',
-				message: 'Which type of device?',
-				choices: [
-					'Photon',
-					'Core',
-					'P1',
-					'Electron'
-				]
-			}]).then((ans) => {
-				const spec = _.find(specs, { productName: ans.type });
-				const binary = spec && spec.knownApps[filePath];
-
-				if (!binary) {
-					throw new VError(`I don't have a ${filePath} binary for ${ans.type}.`);
-				}
-
-				return { map: { binary: binary } };
+				// TODO (mirande): fix this
+				// TODO: this shouldn't be necessary. Just look at the attrs.platform_id instead of
+				// product_id above
+				const question = {
+					name: 'type',
+					type: 'list',
+					message: 'Which type of device?',
+					choices: ['Photon', 'Core', 'P1', 'Electron']
+				};
+
+				return prompt(question)
+					.then(({ type }) => {
+						const spec = _.find(specs, { productName: type });
+						const binary = spec && spec.knownApps[filePath];
+
+						if (!binary){
+							throw new VError(`I don't have a ${filePath} binary for ${type}.`);
+						}
+
+						return { map: { binary: binary } };
+					});
+			})
+			.then((fileMapping) => {
+				return this._doFlash({ product, deviceId, fileMapping });
 			});
-		}).then((fileMapping) => {
-			return this._doFlash({ api, deviceId, fileMapping });
-		});
 	}
 
-	_getDownloadPath(deviceType, saveTo) {
-		if (saveTo) {
+	_getDownloadPath(deviceType, saveTo){
+		if (saveTo){
 			return saveTo;
 		}
 
 		return deviceType + '_firmware_' + Date.now() + '.bin';
 	}
 
-	compileCode(deviceType, files, { target, saveTo, followSymlinks }) {
-		let api;
+	compileCode({ target, followSymlinks, saveTo, params: { deviceType, files } }){
 		let platformId;
 		let targetVersion;
 
-		return Promise.resolve().then(() => {
-			if (files.length === 0) {
-				files.push('.'); // default to current directory
-			}
+		return Promise.resolve()
+			.then(() => ensureAPIToken())
+			.then(() => {
+				if (files.length === 0){
+					files.push('.'); // default to current directory
+				}
 
-			if (deviceType in PLATFORMS) {
-				platformId = PLATFORMS[deviceType];
-			} else {
-				throw new Error([
-					'Target device ' + deviceType + ' is not valid',
-					'	eg. particle compile core xxx',
-					'	eg. particle compile photon xxx'
-				].join('\n'));
-			}
+				if (deviceType in PLATFORMS){
+					platformId = PLATFORMS[deviceType];
+				} else {
+					throw new Error([
+						`Target device ${deviceType} is not valid`,
+						'	eg. particle compile core xxx',
+						'	eg. particle compile photon xxx'
+					].join('\n'));
+				}
 
-			api = new ApiClient();
-			api.ensureToken();
+				this.ui.stdout.write(`${os.EOL}Compiling code for ${deviceType}${os.EOL}`);
 
-			console.log('\nCompiling code for ' + deviceType);
+				if (target){
+					if (target === 'latest'){
+						return;
+					}
 
-			if (target) {
-				if (target === 'latest') {
-					return;
+					return createAPI().listBuildTargets(true /* onlyFeatured */)
```

Comment:
I really dislike that the CLI checks the list of build targets before compiling since it prevents compiling against hidden build targets. It's not a change for now but we should fix this when we rework compiling for LTS.

Reply:
agree 🤝:+1: